src/queues/ThrottlingQueue.ts

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { extractRequestError, LogLevel } from "matrix-bot-sdk";
+import { Mjolnir } from "../Mjolnir";
+
+export type Task<T> = (queue: ThrottlingQueue) => Promise<T>;
+
+/**
+ * A queue for backgrounding tasks without hammering servers too much.
+ */
+export class ThrottlingQueue {
+    /**
+     * The pending tasks.
+     */
+    private tasks: (() => Promise<void>)[] | null;
+
+    /**
+     * A timeout for the next task to execute.
+     */
+    private timeout: ReturnType<typeof setTimeout> | null;
+
+    /**
+     * How long we should wait between the completion of a tasks and the start of the next task.
+     * Any >=0 number is good.
+     */
+    private _delayMS: number;
+
+    /**
+     * Construct an empty queue.
+     *
+     * This queue will start executing whenever `push()` is called and stop
+     * whenever it runs out of tasks to execute.
+     *
+     * @param delayMS The default delay between executing two tasks, in ms.
+     */
+    constructor(private mjolnir: Mjolnir, delayMS: number) {
+        this.timeout = null;
+        this.delayMS = delayMS;
+        this.tasks = [];
+    }
+
+    /**
+     * Stop the queue, make sure we can never use it again.
+     */
+    public dispose() {
+        this.stop();
+        this.tasks = null;
```

Comment:
is this the best way of making sure it can never be used again?

Reply:
I think so?
What else would you suggest? A boolean `isDisposed`?

Reply:
possibly, I am worrying about this only so that it is explicit why the queue is "broken" rather than getting an error because the task queue is full.